src/fs/rock/RockIoState.cc

```diff
@@ -245,44 +255,38 @@ Rock::IoState::writeToBuffer(char const *buf, size_t size)
 }
 
 /// write what was buffered during write() calls
-/// negative sidNext means this is the last write request for this entry
 void
-Rock::IoState::writeToDisk(const SlotId sidNextProposal)
+Rock::IoState::writeToDisk()
 {
     assert(theFile != NULL);
     assert(theBuf.size >= sizeof(DbCellHeader));
 
-    const bool lastWrite = sidNextProposal < 0;
+    assert((sidFirst < 0) == (sidCurrent < 0));
+    if (sidFirst < 0) // this is the first disk write
+        sidCurrent = sidFirst = dir->reserveSlotForWriting();
+
+    // negative sidNext means this is the last write request for this entry
+    const bool lastWrite = sidNext < 0;
+    // here, eof means that we are writing the right-most entry slot
     const bool eof = lastWrite &&
                      // either not updating or the updating reader has loaded everything
                      (touchingStoreEntry() || staleSplicingPointNext < 0);
-    // approve sidNextProposal unless _updating_ the last slot
-    const SlotId sidNext = (!touchingStoreEntry() && lastWrite) ?
-                           staleSplicingPointNext : sidNextProposal;
-    debugs(79, 5, "sidNext:" << sidNextProposal << "=>" << sidNext << " eof=" << eof);
+    debugs(79, 5, "sidCurrent=" << sidCurrent << " sidNext=" << sidNext << " eof=" << eof);
 
     // TODO: if DiskIO module is mmap-based, we should be writing whole pages
     // to avoid triggering read-page;new_head+old_tail;write-page overheads
 
-    writeBufToDisk(sidNext, eof, lastWrite);
-    theBuf.clear();
-
-    sidCurrent = sidNext;
-}
-
-/// creates and submits a request to write current slot buffer to disk
-/// eof is true if and only this is the last slot
-void
-Rock::IoState::writeBufToDisk(const SlotId sidNext, const bool eof, const bool lastWrite)
-{
-    // no slots after the last/eof slot (but partial slots may have a nil next)
-    assert(!eof || sidNext < 0);
+    assert(eof == lastWrite); // no slots after eof
 
     // finalize db cell header
     DbCellHeader header;
     memcpy(header.key, e->key, sizeof(header.key));
-    header.firstSlot = writeAnchor().start;
-    header.nextSlot = sidNext;
+    header.firstSlot = sidFirst;
+
+    const auto updatingTheLastSlot = !touchingStoreEntry() && lastWrite;
```

Comment:
This constant name is confusing (difficult to interpret correctly). The initialization expression means "writing the last slot that should be written when updating the entry". It does not mean "writing the last slot that the entry has". We should fix that:

```suggestion
    const auto lastUpdatingWrite = lastWrite && !touchingStoreEntry();
```

I would also remove the assertion as clearly implied by the initialization code immediately above it. It does not add any value.

At the end, we would have:

```C++
    const auto lastUpdatingWrite = lastWrite && !touchingStoreEntry();
    header.nextSlot = lastUpdatingWrite ? staleSplicingPointNext : sidNext;
```

Reply:
> I would also remove the assertion as clearly implied by the initialization code immediately above it. It does not add any value.

Ah, that is not my/original assertion! That assertion was more meaningful (i.e. not obviously true). It does not mean it was correct, of course! Reviewing that original assertion probably falls into the eof/lastWrite discussion we are having on a different thread in this review, but if there are no discoveries there, then the assertion should stay, and the end result would look like this:

```C++
const auto lastUpdatingWrite = lastWrite && !touchingStoreEntry();
assert(!lastUpdatingWrite || sidNext < 0);
header.nextSlot = lastUpdatingWrite ? staleSplicingPointNext : sidNext;
```

Reply:
Done.